src/time/time.h

```diff
@@ -1,95 +1,114 @@
+ * to time_update().
  */
-rel_time_t now;
+extern rel_time_t now_sec;
+extern rel_time_fine_t now_ms;
+extern rel_time_fine_t now_us;
+extern rel_time_fine_t now_ns;
 
-/* Get the time the process started */
+/* Get the time the process was started */
 static inline time_t
 time_started(void)
 {
-    return time_start;
+    return __atomic_load_n(&time_start, __ATOMIC_RELAXED);
+}
+
+/* Get the current time in seconds since the process started */
+static inline rel_time_t
+time_now(void)
+{
+    return __atomic_load_n(&now_sec, __ATOMIC_RELAXED);
+}
+
+/* Get the current time in milliseconds since the process started */
+static inline rel_time_fine_t
+time_now_ms(void)
+{
+    return __atomic_load_n(&now_ms, __ATOMIC_RELAXED);
+}
+
+/* Get the current time in microseconds since the process started */
+static inline rel_time_fine_t
+time_now_us(void)
+{
+    return __atomic_load_n(&now_us, __ATOMIC_RELAXED);
 }
 
-/* Get the current absolute time (not time since process began) */
+/* Get the current time in nanoseconds since the process started */
+static inline rel_time_fine_t
+time_now_ns(void)
+{
+    return __atomic_load_n(&now_ns, __ATOMIC_RELAXED);
+}
+
+/* Get the current absolute time (not since process began) */
 static inline time_t
 time_now_abs(void)
 {
-    return time_start + (time_t)now;
+    return time_started() + time_now();
 }
 
-/* Get the current time (since process started) */
+/*
+ * time_reltime
+ *
+ * Depending on build options, time_reltime will take on different behaviors.
+ *
+ * In absolute timestamp only mode, t is treated as an absolute unix timestamp,
+ * and time_reltime will return the difference between the timestamp and the
+ * timestamp of when the server came up.
+ *
+ * In relative timestamp only mode, t is treated as a time relative to time now,
+ * and is converted into time since process start.
+ *
+ * In memcached compatibility mode, t is treated as it is in memcached, that is,
+ * if it is greater than 30 days, it is treated as an absolute timestamp;
+ * otherwise, it is treated as a relative timestamp.
+ */
+#if defined TIME_ABSOLUTE
+/*
+ * Get time relative to process start, where t is an absolute unix timestamp.
+ */
 static inline rel_time_t
-time_now(void)
+time_reltime(uint32_t t)
 {
-    return now;
+    return (rel_time_t)(t - time_started());
 }
-
-/* Get time relative to process start given absolute time */
+#elif defined TIME_RELATIVE
+/*
+ * Get time relative to process start, where t is number of seconds from now.
+ */
+static inline rel_time_t
+time_reltime(uint32_t t)
+{
+    return (rel_time_t)t + time_now();
+}
+#elif defined TIME_MEMCACHE
+#define TIME_MAXDELTA (time_t)(60 * 60 * 24 * 30)
+/*
+ * Get time relative to process start given absolute time, memcached style.
+ * If a timestamp is greater than 30 days, then it is treated as an absolute,
+ * unix timestamp. Otherwise, it is treated as relative to now.
+ */
 static inline rel_time_t
 time_reltime(uint32_t t)
```

Comment:
Here we can call it `time_memcache2process_sec()`

For all the above, if the name seems way too long, we can use something else, or shorten them to be `time_u2p_*`, `time_d2p_*` and `time_m2p_sec()` (no need for finer-grain timestamp for memcache)